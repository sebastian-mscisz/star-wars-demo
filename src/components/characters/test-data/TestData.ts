import { Characters, Character, Species } from 'src/common';

export const TestCharacter: Character = {
    name: 'Character1',
    url: 'https://swapi.info/api/people/1',
    homeworld: 'https://swapi.info/api/planets/1',
    species: ['species/1'],
    vehicles: ['https://swapi.info/api/vehicles/1'],
};

export const TestCharacters: Characters = [
    TestCharacter,
    {
        name: 'Character2',
        url: 'https://swapi.info/api/people/2',
        vehicles: ['https://swapi.info/api/vehicles/1', 'https://swapi.info/api/vehicles/2'],
        homeworld: 'https://swapi.info/api/planets/2',
        species: ['species/2'],
    },
    {
        name: 'Character3',
        url: 'https://swapi.info/api/people/3',
        vehicles: [],
        homeworld: 'https://swapi.info/api/planets/3',
        species: ['species/3'],
    },
    {
        name: 'Character4',
        url: 'https://swapi.info/api/people/4',
        vehicles: ['https://swapi.info/api/vehicles/3'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/4'],
    },
    {
        name: 'Character5',
        url: 'https://swapi.info/api/people/5',
        vehicles: ['https://swapi.info/api/vehicles/4', 'https://swapi.info/api/vehicles/5'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/5'],
    },
];

export const TestSpecies: Species = [
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species3',
        url: 'species/3',
    },
    {
        name: 'Species4',
        url: 'species/4',
    },
    {
        name: 'Species5',
        url: 'species/5',
    },
];
