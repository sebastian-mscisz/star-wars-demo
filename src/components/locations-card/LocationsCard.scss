@import 'src/variables';

.sw-locations-card {
    width: calc((100% / var(--columns)) - var(--gap) + (var(--gap) / var(--columns)));

    &__tile {
        background-color: $base-background-color-light;
        border-radius: 25% 25% 100% 100%;
        border: 2px solid $base-color;

        &__link {
            display: flex;
            flex-wrap: wrap;
            text-decoration: none;

            &__image-wrapper {
                width: 100%;
                border-radius: 50%;
                overflow: hidden;
                transition: all 0.3s linear;
                border: 2px solid $base-color;
                aspect-ratio: 1;
                background-color: $base-background-color;

                &__display-spinner {
                    height: 100%;
                    display: flex;
                    justify-content: center;
                }

                &__hide-spinner {
                    display: none;
                }

                &__hide-img {
                    visibility: hidden;
                }

                & img {
                    width: 100%;
                    height: 100%;
                    transition: all 1s linear;
                }
            }

            &__text {
                font-size: 1.4rem;
                width: 100%;
                text-align: center;
                color: $base-color;
                transition: all 0.3s linear;

                &:hover {
                    color: $base-color-light;
                }
            }
        }
    }

    &:hover &__tile__link {
        &__image-wrapper {
            border: 2px solid $base-color-light;

            & img {
                transform: rotate(45deg);
            }
        }
    }

    &:hover &__tile__link__text {
        color: $base-color-light;
    }
}