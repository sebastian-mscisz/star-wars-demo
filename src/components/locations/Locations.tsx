import { withHandlers } from 'src/hoc';
import { LocationsUi } from './ui/LocationsUi';
import { useLocations } from 'src/hooks';

const LocationsWithHandlers = withHandlers(LocationsUi);

export const Locations: React.FC = () => {
    const { data: locations, isLoading, loadingError } = useLocations();

    return <LocationsWithHandlers isLoading={isLoading} locations={locations} hasError={!!loadingError} />;
};
