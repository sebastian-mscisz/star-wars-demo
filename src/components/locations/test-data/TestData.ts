import { Locations, Location } from 'src/common';

export const TestLocation: Location = {
    name: 'Location1',
    url: 'https://swapi.info/api/planets/1',
    residents: ['https://swapi.info/api/people/1'],
    population: '1000',
};

export const TestLocations: Locations = [
    TestLocation,
    {
        name: 'Location2',
        url: 'https://swapi.info/api/planets/2',
        residents: ['https://swapi.info/api/people/2'],
        population: '5343',
    },
    {
        name: 'Location3',
        url: 'https://swapi.info/api/planets/3',
        residents: ['https://swapi.info/api/people/3'],
        population: '13333',
    },
    {
        name: 'Location4',
        url: 'https://swapi.info/api/planets/4',
        residents: ['https://swapi.info/api/people/4', 'https://swapi.info/api/people/5'],
        population: '9999',
    },
    {
        name: 'Location5',
        url: 'https://swapi.info/api/planets/5',
        residents: [],
        population: '999',
    },
];
