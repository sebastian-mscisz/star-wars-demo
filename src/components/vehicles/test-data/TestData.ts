import { Vehicles, Vehicle } from 'src/common';

export const TestVehicle: Vehicle = {
    name: 'Vehicle1',
    url: 'https://swapi.info/api/vehicles/1',
    vehicle_class: 'Air speeder',
    pilots: ['character/1', 'character/2'],
};

export const TestVehicles: Vehicles = [
    TestVehicle,
    {
        name: 'Vehicle2',
        url: 'https://swapi.info/api/vehicles/2',
        pilots: ['character/2'],
        vehicle_class: 'Land speeder',
    },
    {
        name: 'Vehicle3',
        url: 'https://swapi.info/api/vehicles/3',
        pilots: ['character/4'],
        vehicle_class: 'Water speeder',
    },
    {
        name: 'Vehicle4',
        url: 'https://swapi.info/api/vehicles/4',
        pilots: ['character/5'],
        vehicle_class: 'Fire speeder',
    },
    {
        name: 'Vehicle5',
        url: 'https://swapi.info/api/vehicles/5',
        pilots: ['character/5'],
        vehicle_class: 'Air speeder',
    },
];
