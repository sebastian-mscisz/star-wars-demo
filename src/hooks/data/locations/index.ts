export * from './useLocationDetails';
export * from './useLocations';
