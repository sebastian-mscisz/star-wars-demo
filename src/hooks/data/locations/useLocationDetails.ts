import { useMemo } from 'react';
import { LocationWithNeighbours } from 'src/common';
import { TDataState, LocationsDataSelector, fetchLocations, useGetApiSingleData } from 'src/store';
import { findItemByIdWithNeighbors } from '../../utils';

export const useLocationDetails = (locationId: string): TDataState<LocationWithNeighbours | undefined> => {
    const locationsFilter = useMemo(
        () => findItemByIdWithNeighbors('https://swapi.info/api/planets/' + locationId),
        [locationId],
    );

    return useGetApiSingleData(fetchLocations, LocationsDataSelector, locationsFilter);
};
