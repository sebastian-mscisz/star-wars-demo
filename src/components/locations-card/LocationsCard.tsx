import React from 'react';
import { Location, getIdFromUrl, locationDetailsPagePath } from 'src/common';
import { NavLink } from 'react-router-dom';
import './LocationsCard.scss';

type LocationsCardProps = {
    item?: Location;
};

export const LocationsCard: React.ElementType<LocationsCardProps> = ({ item }) => {
    if (!item) return null;

    const { name, url } = item;

    return (
        <div className="sw-locations-card">
            <div className="sw-locations-card__tile">
                <NavLink
                    className={'sw-locations-card__tile__link'}
                    key={name}
                    to={locationDetailsPagePath(getIdFromUrl(url))}
                >
                    <h5 className="sw-locations-card__tile__link__text">{name}</h5>
                    <div className="sw-locations-card__tile__link__image-wrapper">
                        <img src={`./locations/${getIdFromUrl(url)}.jpg`} alt="Star wars planet" />
                    </div>
                </NavLink>
            </div>
        </div>
    );
};
