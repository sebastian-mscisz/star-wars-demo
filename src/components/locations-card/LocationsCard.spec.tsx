import { BrowserRouter } from 'react-router-dom';
import { LocationsCard } from './LocationsCard';
import { render } from '@testing-library/react';
import { TestLocation } from '../locations/test-data';

describe('Locations card', () => {
    it('should render', () => {
        const { asFragment } = render(<LocationsCard item={TestLocation} />, { wrapper: BrowserRouter });
        expect(asFragment()).toMatchSnapshot();
    });
});
