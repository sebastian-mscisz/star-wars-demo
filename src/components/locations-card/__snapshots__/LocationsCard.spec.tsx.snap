// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locations card should render 1`] = `
<DocumentFragment>
  <div
    class="sw-locations-card"
  >
    <div
      class="sw-locations-card__tile"
    >
      <a
        class="sw-locations-card__tile__link"
        href="/location/1"
      >
        <h5
          class="sw-locations-card__tile__link__text"
        >
          Location1
        </h5>
        <div
          class="sw-locations-card__tile__link__image-wrapper"
        >
          <img
            alt="Star wars planet"
            src="/locations/1.jpg"
          />
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
